Feature/integrate notifications

maismin opened the pull request:
- [x] Notifications are pulled from the database when a new product is added or a lower price is found

- [x] Dismiss notifications

maismin committed:
Add notification actions and reducer

- to get notifications from the backend.

maismin committed:
Add ability to remove notifications on logout

maismin committed:
Fix App, Notifications, and SignIn component

- The App and SignIn components retrieve the user's notifications when the user
first visits or logs into the website.

- The Notifications component connects to the Redux store and the green
light shows up when the user has 1 or more notifications.

maismin committed:
Fix seeder.js

- Use Promise.all to load and delete all the files in parallel.

maismin committed:
Integrate notification retrieval

- The user receives a ping via websockets when the a successful product
has been added to the database or if a lower price was found.

maismin committed:
Refactor ShoppingLists and Dashboard component

- Move the ListDetailsDialog and AddItemDialog components from the
ShoppingLists to the Dashboard as the dialogs are not part of the
shopping lists.

maismin committed:
Fix notification updates

- Before, the entire browser refreshed when receiving a new
notification. This was because the entire App would show a circular
progress when the getNotification() request was tracked. Upon removing
this loading state from the component, the browser no longer the entire
dashboard when receiving new notifications. If a user has dialogs open,
those dialogs will remain open instead of being closed.

Esodey commented on client/src/App.js:
```diff
@@ -38,10 +38,15 @@ const styles = () => ({
   useEffect(() => {
-    loadUser();
-  }, [loadUser]);
+    async function load() {
```
custom hook right?

Esodey commented on client/src/components/Dashboard/ShoppingLists.js:
```diff
@@ -72,29 +70,17 @@ function mapListsToCards(shoppingLists, classes, handleOpenList) {
-
-  // eslint-disable-next-line no-unused-vars, no-shadow
-  const { getLists, clearErrors, error, lists, loading } = props;
+  const {
```
Lot of variables. Makes sense

Esodey commented on client/src/pages/Dashboard.js:
```diff
@@ -14,12 +16,43 @@ const useStyles = makeStyles({
   const classes = useStyles();
+  const [openList, setOpenList] = useState(false);
+  const [openAddItem, setOpenAddItem] = useState(false);
+  const [listToDisplay, setListToDisplay] = useState(null);
```
Good renaming

Esodey commented on client/src/components/SignIn/SignInForm.js:
```diff
@@ -57,14 +59,15 @@ function SignInForm({
   }
 
-  const handleSubmit = event => {
+  const handleSubmit = async event => {
```
async makes complete sense

Esodey commented on client/src/components/SocketClient.js:
```diff
@@ -18,20 +21,24 @@ function SocketClient({ isAuthenticated }) {
     socket.emit('authenticate');
```
Still need to look into how these sockets work

Esodey commented on client/src/pages/Dashboard.js:
```diff
@@ -14,12 +16,43 @@ const useStyles = makeStyles({
+  const [openAddItem, setOpenAddItem] = useState(false);
+  const [listToDisplay, setListToDisplay] = useState(null);
+
+  const handleClickOpenList = list => {
```
I can hopefully use this logic for the notifications/product details dialog as well

Esodey commented on server/app/_data/seeder.js:
```diff
@@ -33,9 +35,11 @@ const users = JSON.parse(fs.readFileSync(`${__dirname}/users.json`, 'utf-8'));
-    await List.create(lists);
-    await Product.create(products);
-    await User.create(users);
+    await Promise.all([
```
What is the difference here when you add Promise.all?

Esodey reviewed (approved):
Looks great. Well done!

maismin committed:
Fix getNotifications() action

- Fix browser crash when an unauthorized user visits the website.

maismin committed:
Fix getNotifications() action

- Fix error payload

vic2019 commented on client/src/reducers/notifications.js:
```diff
@@ -12,6 +15,8 @@ export default function(state = initialState, action) {
         notifications: payload.notifications,
       };
+    case CLEAR_NOTIFICATIONS:
+      return initialState;
```
This reminded me that I need to implement clean-up after user logout too..

maismin committed:
Fix issue with lists not being updated

- Fix issue with lists not being updated when new notifications were
received.

maismin committed:
Add functionality to dismiss notifications on FE

- Add action and action types for dismissing notifications.

- Add to notification reducer the ability to change remove dismissed
notifications.

- Fix documentation on the API for dismissing notifications.

- Add button for the user to dismiss the notifications.
